Frameworks/Foundation/NSURL.mm

```diff
@@ -890,4 +908,63 @@ - (BOOL)getPromisedItemResourceValue:(id _Nullable*)value forKey:(NSString*)key
     return StubReturn();
 }
 
+/**
+ @Status Interoperable
+*/
++ (instancetype)URLWithStorageFile:(RTObject<WSIStorageFile>*)storageFile {
+    return [[[self alloc] initWithStorageFile:storageFile] autorelease];
+}
+
+/**
+ @Status Interoperable
+*/
+- (instancetype)initWithStorageFile:(RTObject<WSIStorageFile>*)storageFile {
+    ComPtr<IInspectable> comPtr = [storageFile comObj];
+    ComPtr<IStorageItem> comStorageItem;
+    RETURN_NULL_IF_FAILED(comPtr.As(&comStorageItem));
+
+    // Now cache the storage item so that it can later be retreived.
+    ComPtr<IStorageApplicationPermissionsStatics> storageApplicationPermissionsStatics;
+    RETURN_NULL_IF_FAILED(
+        GetActivationFactory(Wrappers::HStringReference(RuntimeClass_Windows_Storage_AccessCache_StorageApplicationPermissions).Get(),
+                             &storageApplicationPermissionsStatics));
+
+    ComPtr<IStorageItemAccessList> accessList;
+    RETURN_NULL_IF_FAILED(storageApplicationPermissionsStatics->get_FutureAccessList(&accessList));
+
+    // See how close we are to the cap.
+    uint32_t maxItems = 0;
+    RETURN_NULL_IF_FAILED(accessList->get_MaximumItemsAllowed(&maxItems));
+
+    // Documentation says 1000 but lets just make sure it was somthing valid.
+    RETURN_NULL_IF_FALSE(maxItems > 0);
+
+    uint32_t count = 0;
+    Wrappers::HString token;
+
+    static std::mutex s_lock;
+    {
+        std::lock_guard<std::mutex> guard(s_lock);
+
+        // protect the count check and addint logic with a lock since we don't want to race to add the last guy.
+        ComPtr<IVectorView<AccessListEntry>> entries;
+        RETURN_NULL_IF_FAILED(accessList->get_Entries(&entries));
+        RETURN_NULL_IF_FAILED(entries->get_Size(&count));
+
+        if (count >= maxItems) {
+            // All full so delete the oldest one. Though not mentioned, the *last* item in the
+            // list is oldest as sorted by the time added to cache (doesn't update for accesses).
+            AccessListEntry lastEntry{};
+            RETURN_NULL_IF_FAILED(entries->GetAt(count - 1, &lastEntry));
+            RETURN_NULL_IF_FAILED(accessList->Remove(lastEntry.Token));
+        }
+
+        RETURN_NULL_IF_FAILED(accessList->AddOverloadDefaultMetadata(comStorageItem.Get(), token.GetAddressOf()));
+    }
+
+    // now that a token has been acquired, form a file path URL similar to a file reference URL based on it.
+    return [self initFileURLWithPath:[NSString stringWithFormat:@"/.file/id=%@", [NSString _stringWithHSTRING:token.Get()]]
+                         isDirectory:NO];
+}
```

Comment:
Yep I totally agree. It would be super grand to have a clear, coherent design here. Unfortunately the public apis for this force our hand a bit. The call pattern in specific that we are seeing for this scenario is:

UIImage *image = ... ;
      if (!image && [imageURL isFileURL]) {
       image = [UIImage imageWithContentsOfFile:[imageURL path]];
      }

Which means the URL needs to return true for isFileURL and needs to interact withimageWithContentsOfFile when using just the raw path. The way to get this to interop easiest and most fully seemed to be to go ahead and use a real file URL with a custom path. Given that /.file/id=... URLs exist on the reference platform, this seemed sort of natural anyway but I definitely agree this isn't a clean design like I normally advocate for. :-(